### PR TITLE
add duration to build complete message

### PR DESCRIPTION
**Pull request by tomerd**
motivation: give users more information

changes: record build duration and include it in CLI output

### CODE REVIEW

**Comment by tomerd** [comment]
UX:

```
~/code/swift/swift-tools-support-core ± ● main
❯ ~/code/swift/swift-package-manager/.build/debug/swift-package clean
~/code/swift/swift-tools-support-core ± ● main
❯ ~/code/swift/swift-package-manager/.build/debug/swift-build
Building for debugging...
[101/101] Merging module TSCTestSupport
Build complete in 19.20s!
~/code/swift/swift-tools-support-core ± ● main
❯ ~/code/swift/swift-package-manager/.build/debug/swift-build
Building for debugging...
Build complete in 0.11s!
```

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/Build/BuildOperationBuildSystemDelegateHandler.swift [hunk 1]
```diff
@@ -627,12 +627,12 @@ final class BuildOperationBuildSystemDelegateHandler: LLBuildBuildSystemDelegate
...
     }
 
-    func buildComplete(success: Bool) {
+    func buildComplete(success: Bool, duration: DispatchTimeInterval) {
         queue.sync {
             self.progressAnimation.complete(success: success)
             if success {
                 self.progressAnimation.clear()
-                self.outputStream <<< "Build complete!\n"
+                self.outputStream <<< "Build complete in \(duration.descriptionInSeconds)!\n"
```
Comment (tomerd) [inline Sources/Build/BuildOperationBuildSystemDelegateHandler.swift:635]: we could consider not printing anything if the build is < 1s

File: Sources/Build/BuildOperation.swift [hunk 2]
```diff
@@ -135,13 +136,17 @@ public final class BuildOperation: PackageStructureDelegate, SPMBuildCore.BuildS
...
+        let buildStartTime = DispatchTime.now()
+
         buildSystemDelegate?.buildStart(configuration: self.buildParameters.configuration)
 
         // Perform the build.
         let llbuildTarget = try computeLLBuildTargetName(for: subset)
         let success = buildSystem.build(target: llbuildTarget)
 
-        buildSystemDelegate?.buildComplete(success: success)
+        let duration = buildStartTime.distance(to: .now())
```
Comment (WowbaggersLiquidLunch) [inline Sources/Build/BuildOperation.swift:147]: Half observation, half question:

Using `.now` includes time when the system sleeps. What if `.uptimeNanoseconds` is used instead, which excludes time when the system sleeps?

I don't know which is more sensible to the user. It's good to think about whether sleep time should be included though imo.

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
On its own I would find this confusing / misleading, because I don't think it is very clear that we are only talking about the actual build here. Typically users would care how long `swift build` takes as a whole, not some (to them) unclearly defined part of it.

If we are planning to emit more timings for other parts of the process, this seems fine, though.

**Comment by tomerd** [comment]
> On its own I would find this confusing / misleading, because I don't think it is very clear that we are only talking about the actual build here. Typically users would care how long swift build takes as a whole, not some (to them) unclearly defined part of it.

this basically measures the 'build' part, in npm they emit something like `compiled in XX.XXms` in the moral equivalent build step

for the whole process the user could use `time swift build`

> If we are planning to emit more timings for other parts of the process, this seems fine, though.

we are emitting durations today for the resolution steps, also separately per package

File: Sources/Build/BuildOperation.swift (hunk 2 above)
Comment (abertelrud) [reply Sources/Build/BuildOperation.swift:147]: It's a good point, though it's unlikely that the system would sleep in the middle of a build.  But that's a good suggestion and might in fact be more correct.

**Review by abertelrud — APPROVED** [review]

**Comment by neonichu** [comment]
I think we should really first define what the parts are that should be measured and where those parts start and stop. 

I would at least think of the llbuild manifest generation to be part of the "build", not just the execution time. At least that's something we consider part of the build time in Xcode.

**Comment by tomerd** [comment]
> I think we should really first define what the parts are that should be measured and where those parts start and stop.
> 
> I would at least think of the llbuild manifest generation to be part of the "build", not just the execution time. At least that's something we consider part of the build time in Xcode.

@neonichu so move the start time to L134 in your mind? that works for me

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
Maybe, I think we should really split the whole process into parts that could get a duration (even if we don't necessarily output it right now) and make sure we can reasonably account for everything in one of those parts so that they add up to 100%. Once we have that, we can figure out which part we want to designate as the "build".

**Comment by tomerd** [comment]
here is an output from building sourcekit-lsp, without resolving any dependencies:

```
~/code/swift/sourcekit-lsp ± ● main
❯ ~/code/swift/swift-package-manager/.build/debug/swift-package clean
~/code/swift/sourcekit-lsp ± ● main
❯ time ~/code/swift/swift-package-manager/.build/debug/swift-build
[1336/1336] Linking sourcekit-lsp
Build complete in 68.34s!
~/code/swift/swift-package-manager/.build/debug/swift-build  622.87s user 131.82s system 1099% cpu 1:08.64 total
```

duration reported by `time` is more or less the one reported in our `Build complete` message. 

wdyt?

**Comment by abertelrud** [comment]
> here is an output from building sourcekit-lsp, without resolving any dependencies:
> 
> ```
> ~/code/swift/sourcekit-lsp ± ● main
> ❯ ~/code/swift/swift-package-manager/.build/debug/swift-package clean
> ~/code/swift/sourcekit-lsp ± ● main
> ❯ time ~/code/swift/swift-package-manager/.build/debug/swift-build
> [1336/1336] Linking sourcekit-lsp
> Build complete in 68.34s!
> ~/code/swift/swift-package-manager/.build/debug/swift-build  622.87s user 131.82s system 1099% cpu 1:08.64 total
> ```
> 
> duration reported by `time` is more or less the one reported in our `Build complete` message.
> 
> wdyt?

This looks good to me!

**Comment by tomerd** [comment]
@swift-ci test

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
```
/home/buildnode/jenkins/workspace/swift-package-manager-self-hosted-Linux-smoke-test/branch-main/tmp/Miscellaneous_TestDiscovery_Async.nUQfLd/Miscellaneous_TestDiscovery_Async/.build/x86_64-unknown-linux-gnu/debug/AsyncPackageTests.derived/AsyncTests.swift:6:23: error: cannot find 'asyncTest' in scope
12:20:54         ("testAsync", asyncTest(testAsync)),
12:20:54                       ^~~~~~~~~
````

this is unexpected failure due to CI infra not using latest toolchain which includes the xctest change on linux

**Comment by tomerd** [comment]
@swift-ci smoke test linux

**Review by abertelrud — APPROVED** [review]

**Pull request merged**